credentials/credentials.go

```diff
@@ -56,6 +56,64 @@ type PerRPCCredentials interface {
 	RequireTransportSecurity() bool
 }
 
+// SecurityLevel defines the protection level on an established channel.
+type SecurityLevel int
+
+const (
+	// SecurityNone indicates a channel is a plaintext channel that does not provide any protection.
+	SecurityNone SecurityLevel = iota
+	// IntegrityOnly indicates a channel only provides integrity protection.
+	IntegrityOnly
+	// PrivacyAndIntegrity indicates a channel provides both privacy and integrity protection.
+	PrivacyAndIntegrity
+)
+
+// MinimumSecurityLevel should be implemented by PerRPCCredentials type struct indicating
+// the minimum protection level a channel should provide in order to communicate
+// a corresponding PerRPCCredentials instance.
+type MinimumSecurityLevel interface {
+	MinimumSecurityLevel() SecurityLevel
+}
+
+// ConnAuthInfo wraps AuthInfo type instance and specifies the security level of
+// established channel.
+type ConnAuthInfo struct {
+	AuthInfo
+	SecurityLevel SecurityLevel
+}
+
+// AuthType returns the type of ConnAuthInfo as a string.
+func (c *ConnAuthInfo) AuthType() string {
+	return "connAuth"
+}
+
+// NewConnAuthInfo creates a new instance of ConnAuthInfo struct.
+func NewConnAuthInfo(authInfo AuthInfo, level SecurityLevel) AuthInfo {
+	connAuthInfo := &ConnAuthInfo{authInfo, level}
+	return connAuthInfo
+}
+
+// EnableSecurityLevelCheck checks if security level examination should be enabled. It will be enabled
+// only if 1) AuthInfo is an instance of ConnAuthInfo and 2) rpcCred implements
+// MinimumSecurityLevel interface.
+func EnableSecurityLevelCheck(authInfo AuthInfo, rpcCred PerRPCCredentials) bool {
```

Comment:
ShouldCheckSecurityLevel  maybe a better name?

Reply:
Hi Jiangtao, we decided not to do a centralized check, but let each PerRPCCredentials to implement its own check in GetRequestMetadata().